[DR-2838] Dataset schema creation - Step 2 and beyond!

lu-c opened the pull request:
[Link to Dataset Schema Builder on dev](https://jade-cl.datarepo-dev.broadinstitute.org/datasets/new)

This PR:
- Re-orders dataset info drop downs
- If changing cloud provider, reset billing profiles that are available
  and reset and selected billing profiles and regions
- Uses nulls and not empty strings for empty values
- Makes description and curators not required
- Makes sure that deleting columns or tables in the JSON viewer properly
  clears selection
- Makes sure that the input of the datatype dropdown shows the formatted
  value
- Scrolls to the top when changing tabs
- When clicking create, opens a job spinner instead just redirecting to the
  datasets page (and when the job finishes, allow the user to go to the
  new dataset)
- Updates tests to take modal into account
- Modifies the SchemaPanel component to reuse between the live edit and the dataset/snapshot schema views

lu-c committed:
[DR-2809] Added journal methods (not ui yet)

lu-c committed:
[DR-2809] Added journal ui

lu-c committed:
[DR-2809] ran an npm install

lu-c committed:
[DR-2810] Enabled snapshot activity tab

lu-c committed:
Merge branch 'develop' of github.com:DataBiosphere/jade-data-repo-ui into DR-2809

lu-c committed:
Making a second develop to test the build errors

lu-c committed:
Testing the github actions failure

lu-c committed:
Merge branch 'develop2' of github.com:DataBiosphere/jade-data-repo-ui into DR-2809

lu-c committed:
Removed minute checks (cause the unit tests would sometimes be "3 minutes ago" instead of a consistent 2. Rounding problems!

lu-c committed:
Merge branch 'develop' of github.com:DataBiosphere/jade-data-repo-ui into DR-2809

lu-c committed:
Removed package lock version increment, removed unused styles from journal entry view

lu-c committed:
Guess the test needs this!

lu-c committed:
The deployment didnt have the correct server methods (was returning a 404),trying the new package version to see if it fixes it.

Updated the types to use the enum instead of a hard coded string!

lu-c committed:
fixing lint, fixing unit test?

lu-c committed:
Updated LightTable so searchString is an optional parameter

lu-c committed:
Updated Lighttable paging on infinitePaging so it will take into account the current rows and disable the next button / paging options

lu-c committed:
fixing lint

lu-c committed:
Updating the description render to have more data, and extract values for display

lu-c committed:
Updating the description render to have more data, and extract values for display

lu-c committed:
Fixing lint

lu-c committed:
Fixing lint

lu-c committed:
[DR-2836]
	- Added root route for the schema builder page.
	- Added component with styling for the schema builder
	- Updated the MUI tab theme so it is universally updated, and we dont need to inject the styles every time
	- Updated components that share the tab theme

lu-c committed:
[DR-2836]
	- feature flag ability

lu-c committed:
fixed lint

lu-c committed:
[DR-2809] Added details for mutation.description

lu-c committed:
[DR-2809] Added details for mutation.description

lu-c committed:
[DR-2809] updated casing for text

lu-c committed:
Merge branch 'DR-2809' of github.com:DataBiosphere/jade-data-repo-ui into DR-2836

lu-c committed:
[DR-2836] slight css changes to align the page more with the other pages in the app

lu-c committed:
[DR-2837] Added step 1 form!

lu-c committed:
[DR-2837] temp work, updating for lint reasons, but also, starting to add access permissions

lu-c committed:
fixing lint

lu-c committed:
merged with develop

lu-c committed:
Added permissions segment

lu-c committed:
Updated values so they align a little closer to the post values

lu-c committed:
no message

lu-c committed:
removed the submit. mulling over the solution for this form thing

lu-c committed:
Fixed forms so it uses the same one, instead of declaring new forms in each sub component

lu-c committed:
fixed lint

lu-c committed:
added maxwidth because it looked weird on big screens

lu-c committed:
stashing these changes so i can do some quick work on develop

lu-c committed:
testing codemirror package

lu-c committed:
Merge branch 'develop' of github.com:DataBiosphere/jade-data-repo-ui into DR-2838

lu-c committed:
Updated builder! Still has bugs.

lu-c committed:
- Fixed bug where clicking a new column wouldnt refresh the values in the details view
- hooked up actions for the up/down arrows
- added checkbox for adding columns to primary keys

lu-c committed:
- Added error message summary
- Added back button
- fixed autocomplete warning

lu-c committed:
- fixed bug where typing in the json editor would sometimes kill the app

lu-c committed:
- attached the dataset schema to the form values

lu-c committed:
Added ability to submit the dataset. Dont worry, it'll always fail. Will ask tomorrow at work what the "defaultProfileId" is supposed to be. Ciao!

lu-c committed:
Merge branch 'develop' of github.com:DataBiosphere/jade-data-repo-ui into DR-2838

lu-c committed:
[DR-2854] added ability to change the billing profile

lu-c committed:
[DR-2838] updating relationship modal

lu-c committed:
Merge branch 'develop' of github.com:DataBiosphere/jade-data-repo-ui into DR-2837

lu-c committed:
[DR-2838] reorganized button lines

lu-c committed:
[DR-2838] updated to have 10,000 character limit

lu-c committed:
[DR-2838] updated to use GCP constant

lu-c committed:
Merge branch 'develop' of github.com:DataBiosphere/jade-data-repo-ui into DR-2838

lu-c committed:
[DR-2837] updated to remove the freeSolo option, updated error messages to have consistent capitalization

lu-c committed:
[DR-2837] added placeholders to fields

lu-c committed:
[DR-2837] Setting value of cloud platform now clears region option

lu-c committed:
[DR-2837] added form validators

lu-c committed:
[DR-2837] consolidated constant

lu-c committed:
[DR-2837] updated with tabpanel component

lu-c committed:
[DR-2837] updated from base

lu-c committed:
[DR-2837] added key!

lu-c committed:
[DR-2837] added key!

lu-c committed:
[DR-2837] added key!

lu-c committed:
[DR-2837] merged with parent

lu-c committed:
[DR-2837] fixing styles

lu-c committed:
[DR-2837] added onSubmit behavior. Now updates names in relationships if they update

lu-c committed:
[DR-2838] Moved dialog opener to the parent component to allow the component to be reused, added default relationship setting, hooked up menu button for adding a relationship directly to a column

lu-c committed:
[DR-2838] removed the javascript object < -- > json object conversion to allow users to copy paste json content!

lu-c committed:
[DR-2838] deleted testing schema, added placeholder for codemirror segment

lu-c committed:
Merge branch 'develop' of github.com:DataBiosphere/jade-data-repo-ui into DR-2837

lu-c committed:
[DR-2837] fixed crashing due to missing s

lu-c committed:
Merge branch 'DR-2837' of github.com:DataBiosphere/jade-data-repo-ui into DR-2838

lu-c committed:
[DR-2838] created common file with repeated styles before someone can tell me to do that :P Fixed bug where relationshipModal would not collapse expanded tables

lu-c committed:
[DR-2838] created common file with repeated styles before someone can tell me to do that :P Fixed bug where relationshipModal would not collapse expanded tables

lu-c committed:
[DR-2838] cleaning up!

lu-c committed:
[DR-2838] cleaning up the uncontrolled warnings in the ui. Fixed bug where typing "enter" will then trigger the form to submit. Not sure why this didnt happen on the informationView of the builder, but it DOES happen for all fo the fields in the builderView.

lu-c committed:
[DR-2838] set default profile value to the first profile after it loads

lu-c committed:
[DR-2838] added tests for creation view and info view!

lu-c committed:
[DR-2838] added a single test for the builder structure

lu-c committed:
[DR-2838] added a single test for the builder structure

lu-c committed:
[DR-2838] I don't know why, but wrapping these TabPanels in the separate component would break the react-hook-form validation. It started working when i changed it back though.

lu-c committed:
[DR-2838] Added unit test with spys to confirm submit has been run!

lu-c committed:
[DR-2838] Added unit test that tests table creation

lu-c committed:
merged with develop

lu-c committed:
Added test for expand/collapse

lu-c committed:
Added test for expand/collapse

lu-c committed:
[DR-2838] Added test for moving tables up and down

lu-c committed:
[DR-2838] lint

lu-c committed:
[DR-2838] added some column unit tests

lu-c committed:
[DR-2838] Added index to make all column names unique, added protection for when the user deletes out of all of the column/table name

lu-c committed:
[DR-2838] will now remove relationships if a relevant column or table is deleted

snf2ye commented on src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx:
```diff
@@ -80,158 +85,220 @@ const styles = (theme: CustomTheme) => ({
+const DatasetSchemaCreationView = withStyles(styles)(
+  ({ classes, dispatch, profiles, history }: IProps) => {
+    const [currentTab, setCurrentTab] = React.useState(0);
+    const changeTab = (_event: any, newCurrentTab: any) => setCurrentTab(newCurrentTab);
```
Could newCurrentTab be typed?

lu-c committed:
[DR-2838] updated table name (now that the tables update in name to have an increment)

lu-c committed:
[DR-2838] fixed table test

lu-c committed:
[DR-2838] added relationship modal tests!!!

lu-c committed:
[DR-2838] fixed lint

s-rubenstein commented on src/components/dataset/schemaCreation/DatasetSchemaCreationView.test.tsx:
```diff
@@ -0,0 +1,1021 @@
+  user: _.cloneDeep(initialUserState),
+};
+
+// const fillValidInfoFields = () => {
```
This file seems to be largely commented out code. Could we remove that? Do you still need it for some reason?

lu-c replied:
These are valid unit tests! I forgot to uncomment them to make them while working locally

s-rubenstein commented on src/components/dataset/schemaCreation/DatasetSchemaCreationView.test.tsx:
```diff
@@ -0,0 +1,1021 @@
+      });
+
+      it('updating column name should update relationship value', () => {
+        cy.get('#schemabuilder-createTable').click();
```
The setup seems to be repeated between these tests, could that be abstracted to a helper?

lu-c replied:
Added a helper method to call these methods!

s-rubenstein commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,803 @@
+  const [datasetSchema, setDatasetSchema] = useState(
+    getValues().schema as DatasetSpecificationModel,
+  );
+  const [selectedTable, setSelectedTable] = useState(-1);
```
Why does this need to default to `-1`? Is that a dummy value to replicate having not started yet? Could undefined be used as a more semantic representation?

lu-c replied:
When I was setting the state to `undefined`, and then changing it, i was getting an error for setting an unset variable - I think this has to stay

s-rubenstein replied:
Interesting! I wonder if that's a typescript thing. Well, sounds fine to me then.

s-rubenstein commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,803 @@
+    expandedTables: {},
+  });
+
+  const [anchorElDetailsMenu, setAnchorElDetailsMenu] = React.useState<null | HTMLElement>(null);
```
Does this work if you delete the null from the function call? I would expect it to default to undefined in that case.

s-rubenstein replied:
Probably the same issue as above, this sounds good.

s-rubenstein commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,803 @@
+  const onJsonViewerChange = React.useCallback(
+    (value: string) => {
+      try {
+        // using these regexes to add / remove quotation marks just because the viewer will then color code
```
Which regexes? Is that value? potentialSchema?

lu-c replied:
Oh shoot, I deleted those regexes

s-rubenstein commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,803 @@
+      try {
+        // using these regexes to add / remove quotation marks just because the viewer will then color code
+        // the keys and values differently. It's easier to read for modifying, but the downside is that
+        // it isnt copy-pastable as ready json.
```
TOL: Could you override the copy behavior to make it ready json, although that could be a bit magic/behind the scenes for users? Or add a copyToClipboard button that does some work for them?

lu-c replied:
i will add a ticket for a copyToClipboard function, to try to keep this pr down (better late than never?)

s-rubenstein commented:
I added a few initial comments, but I have only gotten a bit into it. Is it possible to break this PR up some? I don't know if I can give a solid review with this many changes at once, especially missing context?

Also, would you be able to drop a few screenshots into the PR description? That might help me understand what is going on.

lu-c committed:
[DR-2838] deleted old comment

lu-c committed:
[DR-2838] uncommented unit tests

snf2ye commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,800 @@
+                      );
+                    }
+                    setDatasetSchema(schemaCopy);
+                  }}
```
I think we generally put this sort of function at the top of the file

lu-c replied:
moved!

snf2ye replied:
Whoops, I'm sorry, I did not mean to submit this comment. It looks like you were following a paradigm where only shared functions are at the top of the file and the rest are in line. I think that's fine if you want to revert. I'd rather stay consistent with the rest of the components.

lu-c commented on src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:
```diff
@@ -144,19 +118,33 @@ const DatasetSchemaInformationView = withStyles(styles)(({ classes }: IProps) =>
+          className={clsx(classes.formLabel, { [classes.formLabelError]: errors.defaultProfile })}
         >
-          Terra project*
+          Billing Project*
```
Update this name to Billing Profile

lu-c commented on src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:
```diff
@@ -144,19 +118,33 @@ const DatasetSchemaInformationView = withStyles(styles)(({ classes }: IProps) =>
-              <MenuItem value="yes">Yes</MenuItem>
-              <MenuItem value="no">No</MenuItem>
-            </Select>
+            <Autocomplete
```
Update this to a select

lu-c commented on src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:
```diff
@@ -144,19 +118,33 @@ const DatasetSchemaInformationView = withStyles(styles)(({ classes }: IProps) =>
+            />
           )}
         />
       </Grid>
```
Make default secure monitoring value "no"

lu-c commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,800 @@
+      maxWidth: 200,
+    },
+    jsonWrapper: {
+      backgroundColor: '#272822',
```
add comment describing what this color is

lu-c commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,800 @@
+  } as any);
+
+interface IProps extends WithStyles<typeof styles> {
+  userEmail: string;
```
deletethis

lu-c commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,800 @@
+    expandedTables: {},
+  });
+
+  const [anchorElDetailsMenu, setAnchorElDetailsMenu] = React.useState<null | HTMLElement>(null);
```
Remove `React.`

lu-c commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,800 @@
+                </label>
+                <Autocomplete
+                  id="column-datatype"
+                  options={columnDatatypeOptions}
```
Remove this variable and inline `_.keys()`

lu-c commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,800 @@
+                  size="small"
+                  color="primary"
+                  className={classes.iconButton}
+                  onClick={handleClickDetailsMenu}
```
let's pull out the icon stuff to a thing. :P

lu-c commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,800 @@
+  };
+  const handleCloseDetailsMenu = () => setAnchorElDetailsMenu(null);
+
+  const onJsonViewerChange = React.useCallback(
```
remove `React.`?

lu-c commented:
disable column checkbox when one is selected

lu-c committed:
[DR-2838] Responding to PR feedback!

lu-c committed:
[DR-2838] Responding to PR feedback!

lu-c committed:
[DR-2838] Updated billing profile to select (instead of autocomplete)

lu-c committed:
[DR-2838] Added details menu button helper component

snf2ye commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,780 @@
+                  size="small"
+                  color="primary"
+                  className={classes.iconButton}
+                  disabled={selectedTable === -1}
```
There are other cases when we could disable this, like when the column is the last in the last table, etc. But I wonder if it's worth the more complicated logic.

lu-c replied:
I have a unit test that confirms that clicking the button when the col is already the last in the list doesnt do anything - We could disable the button instead too!

snf2ye commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,780 @@
+                  style={{ marginLeft: 50 }}
+                  disabled={!datasetSchema.tables || datasetSchema.tables.length < 2}
+                  onClick={() => {
+                    setRelationshipModalDefaultValues({
+                      from: '',
+                      to: '',
+                      expandedTables: {},
+                    });
```
This matches the original default value - can we put this into a constant and share it between the two?

snf2ye commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,780 @@
+    });
+  };
+
+  const deleteTable = () => {
```
I mentioned the idea of having a confirmation dialog to confirm before deleting. I don't think I see that yet. 
(Also totally fine to push that off to another PR/ticket)

lu-c replied:
I totally forgot about this! Let me add a ticket for it, and then if I can get it in today, I'll do it

snf2ye reviewed (commented):
Thanks for walking us through some of the code today!
Just a few odds & ends comments.

lu-c committed:
[DR-2838] moved method further up on the class, created default relationship constant, typed function

lu-c committed:
[DR-2832] disabled up down buttons if selected table is at the top of the list

lu-c commented:
Updated the PR with the comments - First, there is definitely a bug, but I could use help looking into it!

Right now, when you click on the detailed menu button, it now shows up like this:
<img width="1429" alt="Screen Shot 2022-12-13 at 7 31 12 PM" src="https://user-images.githubusercontent.com/10501914/207474977-7367f709-03f0-4beb-a090-7cc2548f024e.png">
and it is totally detached from the popping element. I'm not sure exacty what changed to cause this and how to fix - the anchor element is still being declared the same way as before. The only difference is that the IconButton is now being handled in a helper component

Any ideas? :( Otherwise, i will revert this back to duplicating code

lu-c committed:
Merge branch 'develop' of github.com:DataBiosphere/jade-data-repo-ui into DR-2838

lu-c committed:
[DR-2832] fixed disabling logic

lu-c committed:
[DR-2838] Removed details menu button pop up, im sorry, i really am

lu-c committed:
Merge branch 'develop' of github.com:DataBiosphere/jade-data-repo-ui into DR-2838

lu-c committed:
[DR-2838] Added tooltips to dataset schema controls. Disabled fields required / array_of fields based off of back end technical limitations.

lu-c committed:
[DR-2838] fixed test

lu-c committed:
[DR-2838] fixed warnings (tooltip wouldnt work on a disabled button)

lu-c committed:
Merge branch 'develop' of github.com:DataBiosphere/jade-data-repo-ui into DR-2838

lu-c committed:
[DR-2838] fixed lint

lu-c committed:
Merge branch 'develop' of github.com:DataBiosphere/jade-data-repo-ui into DR-2838

lu-c committed:
[DR-2838] Added relationship editor (#1267)

* [DR-2838] Added relationship editor

* [DR-2838] fixed key issue

* [DR-2838] Updated relationship details so it uses the modal instead of the details panel

* [DR-2838] added relationship highlight on hover

* [DR-2838] fixing lint

* [DR-2875] reorganizing the code so it's organized (like things together!)

* [DR-2875] reorganizing the code so it's organized (like things together!)

lu-c committed:
[DR-2838] fixed unit test

lu-c committed:
Merge branch 'DR-2838' of github.com:DataBiosphere/jade-data-repo-ui into DR-2838

nmalfroy committed:
Followon changes

This commit captures a few things:
- Re-ordered dataset info drop downs
- If changing cloud provider, reset billing profiles that are available
  and reset and selected billing profiles and regions
- Use nulls and not empty strings for empty values
- Make descrition and curators not required
- Make sure that deleting columns or tables in the JSON viewer properly
  clears selection
- Make sure that the input of the datatype dropdown shows the formatted
  value
- Scroll to the top when changing tabs
- When clicking create, open a job spinner instead just redirecting to the
  datasets page (and when the job finishes, allow the user to go to the
  new dataset)
- Update tests to take modal into account

nmalfroy committed:
Fix linting

nmalfroy committed:
Merge branch 'develop' into DR-2838

nmalfroy committed:
Fix package lock and vulnerabilities

nmalfroy committed:
Merge branch 'develop' into DR-2838

# Conflicts:
#	package-lock.json

nmalfroy committed:
Update schema creation to use common schema component

snf2ye commented on src/components/common/overview/SchemaPanel.tsx:
```diff
@@ -172,48 +212,120 @@ const renderColumnName = (
+}
+
+interface IPanelProps extends WithStyles<typeof styles> {
+  // Tables to render
+  tables: Array<TableModel>;
+  // If true, will remove hover styling
+  readOnly?: boolean;
+  // If true, render a radio button to represent selection for columns
+  selectedColumnnsAsRadio?: boolean;
+  // Selected node by id where the id is either {table index} if the selected node is a table or {table index}-{column index} if the selected node is a column
+  selected?: string;
+  // If passed in, callback method to execute when any node gets selected
+  onNodeSelect?: (event: React.SyntheticEvent, nodeId: string) => void;
+  // Expanded nodes by id where the id is the {table index}
+  expanded?: Array<string>;
+  // Highlighted nodes by id where the id is the {table index}-{column index}
+  highlighted?: Array<string>;
+  // If passed in, callback method to execute when any node gets toggled
+  onNodeToggle?: (event: React.SyntheticEvent, nodeIds: string[]) => void;
+  // Array of react elements to add before a column label
+  beforeLabelIcons?: React.ReactElement[];
+  // Array of react elements to add after a column label
+  afterLabelIcons?: (table: TableModel, column: ColumnModel) => LabelIcon[];
```
Thanks for the helpful comments!

snf2ye reviewed (approved):
This is awesome! Thank you!

okotsopoulos reviewed (approved):
Thank you for the work on this!

I'm approving because my chief criteria for merging was that users who do *not* enable the feature flag see no degradation of behavior, I was able to interact with dataset and snapshot schemas in your dev environment with no issues in rendering (I know that was shared code that was changed).

I'm going through to offer comments on the new feature now, but didn't want to hold up your merge on them as I know we'll continue to work in this space going forward.

okotsopoulos commented on src/components/dataset/schemaCreation/DatasetCreationModal.tsx:
```diff
@@ -0,0 +1,97 @@
+
+const DatasetCreationModal = withStyles(styles)(
+  ({ classes, isOpen, dataset, creationIsProcessing }: IProps) => {
+    if (creationIsProcessing) {
```
If the dataset creation job fails, it's hard for the user to debug.

As an example, I tried to create a dataset with a billing profile whose account I was missing permissions for.  I got this ephemeral generic error message:
<img width="366" alt="Screen Shot 2023-01-20 at 1 04 38 PM" src="https://user-images.githubusercontent.com/79769153/213792476-500117d1-d686-4f5b-8d4b-d5056243cf3d.png">

But the underlying job has a more descriptive error cause:
<img width="1743" alt="Screen Shot 2023-01-20 at 1 04 00 PM" src="https://user-images.githubusercontent.com/79769153/213792471-8d71b83c-39ae-4228-a5f0-b35231d87335.png">

This is also an issue I've noticed in failed snapshot creation within the UI.  Probably would benefit from some separate design and planning, but it would be nice to provide a link to the job result within the UI.  We currently display job results on click within the Activity tab in a pop-up, but don't link directly (and the Activity tab is also unusable to non-admins given slow performance of Stairway queries).

okotsopoulos commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,953 @@
+    </div>
+  );
+});
+
```
I noticed that we can set relationships in this schema wizard, but we can't set assets.  AFAIK, a dataset must have assets in order to be used in TDR UI's snapshot creation wizard.  I'd recommend that we prioritize supporting their specification here.

okotsopoulos commented on src/components/dataset/schemaCreation/DatasetSchemaInformationView.tsx:
```diff
@@ -208,7 +157,42 @@ const DatasetSchemaInformationView = withStyles(styles)(({ classes }: IProps) =>
+                  placeholder="Default billing profile"
+                />
+              )}
+              getOptionLabel={(option: BillingProfileModel) => option.profileName || ''}
```
I agree with displaying a billing profile by its human-readable name, but what do you think about displaying the UUID secondarily?  Maybe something like
```
name (aaaa-bbbb-cccc-dddd)
```
where the UUID is styled to grey.  I'm thinking in case the user wants to interact with our billing profile endpoints, they'll need the UUID.

Also not sure if it's worth calling out in some way at this stage that even if a billing profile is accessible to the user, they will also need access to its underlying billing account… maybe the dataset creation error message in that case covers our bases well enough.

okotsopoulos commented on src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx:
```diff
@@ -80,58 +86,100 @@ const styles = (theme: CustomTheme) => ({
+      defaultProfileId: null,
+      region: null,
       stewards: [],
       custodians: [],
```
Is there a reason why we don't allow snapshot creators to be set at this stage too?

When we allow the policy members to be set here, is there an opportunity to reuse code from snapshot creation? The difference in UX threw me off a bit. 
<img width="596" alt="Screen Shot 2023-01-20 at 3 15 40 PM" src="https://user-images.githubusercontent.com/79769153/213797556-10ec331a-99ae-466a-9069-1fcebc5c2c10.png">

okotsopoulos commented on src/components/dataset/schemaCreation/DatasetSchemaCreationView.tsx:
```diff
@@ -185,7 +233,25 @@ const DatasetSchemaCreationView = withStyles(styles)(({ classes }: IProps) => {
                   </Button>
                 )}
-              </TabPanel>
+                {_.keys(errors).length > 0 && (
```
Sorry, I don't think this is the best place for this comment!

I noticed that if I introduce an error to the schema definition but fix it, the error message remains:
<img width="872" alt="Screen Shot 2023-01-20 at 12 21 52 PM" src="https://user-images.githubusercontent.com/79769153/213799279-24093047-7aeb-4ce8-99c7-bb0cb0140e1d.png">

I also noticed that you can still submit a dataset creation request even if your form has errors.  Is that intentional?

okotsopoulos commented on src/components/dataset/schemaCreation/DatasetSchemaBuilderView.tsx:
```diff
@@ -0,0 +1,953 @@
+  };
+
+  // ----------------------------------------
+  // CodeMirror (JSON viewer)
```
The CodeMirror is super cool.  I wonder if we want to "lock" it on default, though?  I could see a user wanting to create the schema via the GUI, inadvertently modifying the mirrored JSON, and not being clear on how to fix.  (For example, you could change the JSON for a column definition to illegally stack several types: it would then reflect in the GUI but you'd be unable to fix it there.)

<img width="1477" alt="Screen Shot 2023-01-20 at 12 22 22 PM" src="https://user-images.githubusercontent.com/79769153/213799975-2be77bd2-9945-4f00-9ddf-d58b5d7155db.png">

okotsopoulos reviewed (approved):
As promised, my follow-on comments from interacting with this feature :-)
